examples/usercall-extension/runner/src/main.rs

```diff
@@ -0,0 +1,115 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate aesm_client;
+extern crate enclave_runner;
+extern crate sgxs_loaders;
+
+use aesm_client::AesmClient;
+use enclave_runner::EnclaveBuilder;
+use enclave_runner::usercalls::{UsercallExtension, SyncStream};
+use sgxs_loaders::isgx::Device as IsgxDevice;
+
+use std::io::{Result as IoResult, Write, Read};
+use std::process::{Command, Child, Stdio};
+use std::sync::{Arc, Mutex};
+
+/// This example demonstrates use of user-call extensions.
+/// User call extension allow the enclave code to "connect" to an external service via a customized enclave runner.
+/// Here we customize the runner to intercept calls to connect to an address "cat" which actually connects the enclave application to 
+/// stdin and stdout of `cat` process.
+
+struct CatService {
+   c : Arc<Mutex<Child>>
+}
+impl CatService {
+   fn new() -> Result<CatService, std::io::Error> {
+        Command::new("/bin/cat")
+            .stdout(Stdio::piped())
+            .stdin(Stdio::piped())
+            .spawn()
+            .map(|c| Arc::new(Mutex::new(c)))
+            .map(|c| CatService {c})
+   }
+}
+impl SyncStream for CatService {
+
+    fn read(&self, buf: &mut [u8]) -> IoResult<usize> {
+        self.c.lock().unwrap().stdout.as_mut().unwrap().read(buf)
+    }
+
+    fn write(&self, buf: &[u8]) -> IoResult<usize> {
+        self.c.lock().unwrap().stdin.as_mut().unwrap().write(buf)
+    }
+
+    fn flush(&self) -> IoResult<()> {
+        self.c.lock().unwrap().stdin.as_mut().unwrap().flush()
+    }
+}
+
+#[derive(Debug)]
+struct ExternalService;
+impl UsercallExtension for ExternalService {
+    fn connect_stream(
+        &self,
+        addr: &str,
+        local_addr: Option<&mut String>,
+        peer_addr: Option<&mut String>,
+    ) -> IoResult<Option<Box<SyncStream>>> {
+        // If the passed address is not "cat", we return none, whereby the passed address gets treated as 
+        // an IP address which is the default behavior.
+        match &*addr {
+           "cat" => { 
+               let stream = CatService::new()?;
+               if let Some(local_addr) = local_addr {
+                   local_addr.push_str("None");
+               }
+               if let Some(peer_addr) = peer_addr {
+                   peer_addr.push_str("None");
+               }
```

Comment:
We shouldn't push the string `"None"` here, since it is not a valid address. Rust will error when it tries to parse it. Unless I'm missing something you already discussed with Jethro, just ignore `local_addr` and `peer_addr` unless you want to make use of them for some reason in the enclave.

We could even add another version of `connect_stream` later to this trait that doesn't take `local_addr` and `peer_addr` so that the user doesn't have to think about them at all when not relevant (which I believe will be most of the time).

Reply:
Sure. I will remove the "None" and left a comment that we are ignoring local_addr and peer_addr since it is not relevant to the context of the example.